securesystemslib/hsm.py

```diff
@@ -0,0 +1,446 @@
+# division.  Example:  print 'hello world' raises a 'SyntaxError' exception.
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import unicode_literals
+
+# Import python wrapper for PKCS#11 to communicate with the tokens
+import PyKCS11
+
+import binascii
+import logging
+import securesystemslib.exceptions
+
+# Import cryptography routines needed to retrieve cryptographic
+# keys and certificates in PEM format.
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.backends import default_backend
+from cryptography import x509
+
+# RSA-PSS with SHA256 hash to be used for signature generation.
+RSA_PSS_MECH = PyKCS11.CKM_SHA256_RSA_PKCS_PSS
+# SHA256 hash to be used to digest the data.
+RSA_PSS_HASH_SHA256 = PyKCS11.CKM_SHA256
+# Mask generating function for SHA256 Hash.
+RSA_PSS_MGF_SHA256 = PyKCS11.CKG_MGF1_SHA256
+# Length of salt to be used for hashing.
+RSA_PSS_SALT_LENGTH = 32
+
+logger = logging.getLogger('securesystemslib_hsm')
+
+class HSM(object):
+  """
+  <Purpose>
+    Provides an interface to use cryptographic tokens for various
+    cryptographic operations.
+
+  <Arguments>
+    PKCS11Lib_path:
+       path to the PKCS#11 library. This can be module specific or
+       library by OpenSC(opensc-pkcs11.so) can be used.
+  <Exceptions>
+    securesystemslib.exceptions.NotFoundError, if the path of PKCS#11
+    library is not specified or the library is corrupt
+  """
+
+  def __init__(self, PKCS11Lib_path):
+
+    self.PKCS11LIB = PKCS11Lib_path
+
+    # Initialize the PyKCS11Lib, wrapper of PKCS#11 in Python.
+    self.PKCS11 = PyKCS11.PyKCS11Lib()
+
+    self.session = None
+
+    # Load the PKCS11 shared library file.
+    self.refresh()
+
+
+
+
+
+  def refresh(self):
+    """
+    <Purpose>
+      This method refreshes the list of available cryptographic tokens.
+
+    <Exceptions>
+      securesystemslib.exceptions.NotFoundError, if the path of PKCS#11
+      library is not specified or the library is corrupt
+    """
+
+    # Try to load the PKCS11 library
+    try:
+      # Load the PKCS#11 library and simultaneously update the list
+      # of available HSM.
+      self.PKCS11.load(self.PKCS11LIB)
+    except PyKCS11.PyKCS11Error as error:
+      logger.info('PKS11 Library not found or is corrupt!')
+      raise securesystemslib.exceptions.NotFoundError(error.__str__())
+
+
+
+
+
+  def get_available_HSMs(self):
+    """
+    <Purpose>
+      Generate the list of available cryptographic tokens for the user
+
+    <Exceptions>
+      securesystemslib.exceptions.NotFoundError, if the path of PKCS#11
+      library is not specified or the library is corrupt
+
+    <Returns>
+      A list of dictionaries consisting of relevant information
+      regarding all the available tokens
+    """
+
+    # Refresh the list of available slots for HSM
+    self.refresh()
```

Comment:
This is a more general design question: what is the intended lifetime behavior of `HSM` objects?

My intuition as a consumer of this library is that the list of available HSMs available for use in an `HSM` instance is fixed at initialization time, but calling `refresh` here makes it possible that the list has changed between object creation and use. If the intended behavior is that any operation that needs to access HSMs always operates on a refreshed list, it might make sense to turn `refresh` into a decorator and explicitly document that as the behavior.

Reply:
The intended behaviour of the `HSM` object is something like, get the list of available HSM -> establish a session with the required HSM -> do the intended job or jobs -> close the session.
It's advisable to refresh the list before doing a series of operations as SlotID may change on removal and reinsertion of the HSM.
In `hsm_keys`, `load_HSM` calls `refresh` and get the list of available HSMs to the user. The user chooses from the list of slot available and then pass it as the argument `HSM_info` to various functions. Adding `refresh` as a decorator won't give the user the functionality to choose from the available HSMs.